platforms/posix/src/lockstep_scheduler/src/lockstep_scheduler.cpp

```diff
@@ -1,102 +1,123 @@
 	}
 }
 
 int LockstepScheduler::cond_timedwait(pthread_cond_t *cond, pthread_mutex_t *lock, uint64_t time_us)
 {
-	std::shared_ptr<TimedWait> new_timed_wait;
+	// A TimedWait object might still be in timed_waits_ after we return, so its lifetime needs to be
+	// longer. And using thread_local is more efficient than malloc.
+	static thread_local TimedWait timed_wait;
 	{
-		std::lock_guard<std::mutex> lock_timed_waits(timed_waits_mutex_);
+		std::lock_guard<std::mutex> lock_timed_waits(_timed_waits_mutex);
 
 		// The time has already passed.
-		if (time_us <= time_us_) {
+		if (time_us <= _time_us) {
 			return ETIMEDOUT;
 		}
 
-		new_timed_wait = std::make_shared<TimedWait>();
-		new_timed_wait->time_us = time_us;
-		new_timed_wait->passed_cond = cond;
-		new_timed_wait->passed_lock = lock;
-		timed_waits_.push_back(new_timed_wait);
-		timed_waits_iterator_invalidated_ = true;
+		timed_wait.time_us = time_us;
+		timed_wait.passed_cond = cond;
+		timed_wait.passed_lock = lock;
+		timed_wait.timeout = false;
+		timed_wait.done = false;
+
+		// Add to linked list if not removed yet (otherwise just re-use the object)
+		if (timed_wait.removed) {
+			timed_wait.removed = false;
+			timed_wait.next = _timed_waits;
+			_timed_waits = &timed_wait;
+		}
 	}
 
-	while (true) {
-		int result = pthread_cond_wait(cond, lock);
+	int result = pthread_cond_wait(cond, lock);
 
-		// We need to unlock before aqcuiring the timed_waits_mutex, otherwise
-		// we are at rist of priority inversion.
-		pthread_mutex_unlock(lock);
+	const bool timeout = timed_wait.timeout;
 
-		{
-			std::lock_guard<std::mutex> lock_timed_waits(timed_waits_mutex_);
-
-			if (result == 0 && new_timed_wait->timeout) {
-				result = ETIMEDOUT;
-			}
-
-			new_timed_wait->done = true;
-		}
+	if (result == 0 && timeout) {
+		result = ETIMEDOUT;
+	}
 
-		// The lock needs to be locked on exit of this function
+	timed_wait.done = true;
+
+	if (!timeout && _setting_time) {
+		// This is where it gets tricky: the timeout has not been triggered yet,
+		// and another thread is in set_absolute_time().
+		// If it already passed the 'done' check, it will access the mutex and
+		// the condition variable next. However they might be invalid as soon as we
+		// return here, so we wait until set_absolute_time() is done.
+		// In addition we have to unlock 'lock', otherwise we risk a
+		// deadlock due to a different locking order in set_absolute_time().
+		// Note that this case does not happen too frequently, and thus can be
+		// a bit more expensive.
+		pthread_mutex_unlock(lock);
+		_timed_waits_mutex.lock();
+		_timed_waits_mutex.unlock();
 		pthread_mutex_lock(lock);
-		return result;
 	}
+
+	return result;
 }
 
 int LockstepScheduler::usleep_until(uint64_t time_us)
 {
-	pthread_mutex_t lock;
-	pthread_mutex_init(&lock, nullptr);
-	pthread_cond_t cond;
-	pthread_cond_init(&cond, nullptr);
+	pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;
+	pthread_cond_t cond = PTHREAD_COND_INITIALIZER;
```

Comment:
I thought these were only allowed for `static` locks/conds but looks like that's only true for older versions of the POSIX standard. I guess it's a nice optimization to save calls.